### PR TITLE
Implementation of exact algorithms for minimum dominating set

### PR DESCRIPTION
**Pull request by PiotrKubaty**

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 add_subdirectory(coloring)
+add_subdirectory(dominatingset)
```
Comment (krzysztof-turowski) [inline cpp/CMakeLists.txt:2]: Please git mv both code and includes to "dominating_set" directory

File: cpp/dominatingset/BranchAndReduceMSC.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
+#include <dominatingset/RooijBodlaenderMDS.hpp>
+#include <dominatingset/BranchAndReduceMSC.hpp>
+#include <cassert>
+
+NetworKit::index findCountingRuleReductionSet(std::vector<std::set<NetworKit::node>> &family, std::vector<std::set<NetworKit::index>> &occurences);
+NetworKit::index find2Cardinality2FrequencySet(std::vector<std::set<NetworKit::node>> &family, std::vector<std::set<NetworKit::index>> &occurences);
+
+template <typename T>
+std::set<T> setUnion(const std::set<T>& first, const std::set<T>& second)
+{
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/BranchAndReduceMSC.cpp:10]: Please apply [K&R style](https://en.wikipedia.org/wiki/Indentation_style#K&R_style) braces consistently

File: cpp/dominatingset/BranchAndReduceMSC.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#include <dominatingset/RooijBodlaenderMDS.hpp>
+#include <dominatingset/BranchAndReduceMSC.hpp>
+#include <cassert>
+
+NetworKit::index findCountingRuleReductionSet(std::vector<std::set<NetworKit::node>> &family, std::vector<std::set<NetworKit::index>> &occurences);
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/BranchAndReduceMSC.cpp:5]: Please run the following command and fix linter errors:

cpplint --linelength=100 --extensions=cpp,hpp --filter=-legal/copyright,-build/c++20,-build/include,-build/namespaces,-build/header_guard,-runtime/references,-runtime/string --recursive --exclude=./backup/* --exclude=./build/* --exclude=./input/* --exclude=./lib/* --exclude=./cpp/flow/maximum_flow/dynamic_tree/* .

File: cpp/dominatingset/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+koala_add_module(dominatingset
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/CMakeLists.txt:1]: dominating_set

File: cpp/dominatingset/MinimumDominatingSet.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    for (bool is_dominated : dominated) {
+        if (!is_dominated) {
+            return false;
+        }
+    }
+    return true;
+}
+
+std::vector<bool> &smallerCardinalitySet(std::vector<bool> &lhs, std::vector<bool> &rhs) {
+    // std::min(lhs, rhs, [](std::vector<bool> set) { return std::count(set.begin(), set.end(), true); })
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/MinimumDominatingSet.cpp:24]: Please clean-up all commented code

File: cpp/dominatingset/MinimumDominatingSet.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+                dominated[v] = true;
+            });
+        }
+    });
+    for (bool is_dominated : dominated) {
+        if (!is_dominated) {
+            return false;
+        }
+    }
+    return true;
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/MinimumDominatingSet.cpp:20]: Please use https://en.cppreference.com/w/cpp/algorithm/all_any_none_of

File: cpp/dominatingset/RooijBodlaenderMDS.cpp [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+std::vector<bool> minimumSetCover(std::vector<std::set<NetworKit::node>> &family, std::vector<std::set<NetworKit::index>> &occurences) {
+    bool emptyInstance = true;
+    for (auto &familyElement : family) {
+        if (familyElement.size() != 0) {
+            emptyInstance = false;
+            break;
+        }
+    }
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/RooijBodlaenderMDS.cpp:51]: https://en.cppreference.com/w/cpp/algorithm/all_any_none_of strikes again

File: cpp/dominatingset/SchiermeyerMDS.cpp [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+    }
+    return graph;
+}
+
+NetworKit::Graph core(const NetworKit::Graph &G, std::set<NetworKit::node> &free, std::set<NetworKit::node> &bounded, std::set<NetworKit::node> &required) {
+    std::vector<std::set<NetworKit::node>> intermediate(G.numberOfNodes());
+    
+    G.forNodes([&G, &intermediate](NetworKit::node u) {
+        G.forNeighborsOf(u, [u, &intermediate](NetworKit::node neighbor) {intermediate[u].insert(neighbor);});
+    });
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/SchiermeyerMDS.cpp:58]: G.forEdges is not sufficient?

File: cpp/dominatingset/SchiermeyerMDS.cpp [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    //     graph.forEdgesOf(u, [u](NetworKit::node v) {
+    //         std::cout << v << " ";
+    //     });
+    //     std::cout << "]\n";
+    // });
+    // std::cout << "with ";
+    // for (auto e : S) {
+    //     std::cout << e << " ";
+    // } std::cout << "\n";
+}
```
Comment (krzysztof-turowski) [inline cpp/dominatingset/SchiermeyerMDS.cpp:275]: Well, this has to be cleaned up.

File: test/testDominatingSet.cpp [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+#include <dominatingset/BranchAndReduceMDS.hpp>
+#include <dominatingset/BranchAndReduceMSC.hpp>
+#include <dominatingset/SchiermeyerMDS.hpp>
+#include <dominatingset/FominKratschWoeginger.hpp>
+#include <dominatingset/ExhaustiveMDS.hpp>
+#include <chrono>
+#include <io/G6GraphReader.hpp>
+#include <fstream>
+#include<algorithm>
+#include<iterator>
```
Comment (krzysztof-turowski) [inline test/testDominatingSet.cpp:12]: Once it's #include <A>, another time #include<A>, please unify it with reference to the previous files (e.g. take recognition/Perfect files as a reference)

File: test/testDominatingSet.cpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
+// #include <iostream>
```
Comment (krzysztof-turowski) [inline test/testDominatingSet.cpp:1]: Please rewrite tests so that they are similar in spirit to e.g. testVertexColoring

In general, test is for small/corner cases, benchmark is for large suites (e.g. big g6 input files)

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Thank you very much, the code works well, but still it requires a lot of cleanup to pass the review.

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much for this well-thought and well-implemented contribution. All is good to go!
